PayloadPkg/OsLoader/OsLoader.c

```diff
@@ -644,78 +881,104 @@ BootOsImage (
+    goto ERROR;
   }
 
-  // Load normal OS boot image
-  LoadedImage.HwPartHandle = HwPartHandle;
-  Status = LoadAndSetupImage (&BootOption, &LoadedImage);
+  //
+  // Init File System
+  //
+  Status = InitBootFileSystem (OsBootOption, HwPartHandle, &FsHandle);
   if (EFI_ERROR (Status)) {
-    DEBUG ((DEBUG_INFO, "LoadImage: Status = %r\n", Status));
-    return Status;
+    DEBUG ((DEBUG_INFO, "Failed to Initialize Boot File System - SwPart %d\n", OsBootOption->SwPart));
+    goto ERROR;
   }
 
-  // Load Trusy OS image if required
-  if ((BootOption.BootFlags & BOOT_FLAGS_TRUSTY) != 0) {
-    ZeroMem (&LoadedTrustyImage, sizeof (LOADED_IMAGE));
-    LoadedTrustyImage.HwPartHandle  = HwPartHandle;
-    LoadedTrustyImage.LoadImageType = LOAD_IMAGE_TRUSTY;
-    Status = LoadAndSetupImage (&BootOption, &LoadedTrustyImage);
-    if (EFI_ERROR (Status)) {
-      DEBUG ((DEBUG_INFO, "Load extra Image: Status = %r\n", Status));
-      return Status;
-    }
+  //
+  // Load Boot Image
+  //
+  Status = LoadBootImages (OsBootOption, HwPartHandle, FsHandle, &LoadedImageHandle);
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_INFO, "Failed to Load Boot Image\n"));
+    goto ERROR;
   }
+  AddMeasurePoint (0x4070);
 
-  // Load Extra OS image if required
-  if ((BootOption.BootFlags & BOOT_FLAGS_EXTRA) != 0) {
-    for (Index = 0; Index < MAX_EXTRA_IMAGE_NUM; Index++) {
-      if (BootOption.Image[LOAD_IMAGE_EXTRA0 + Index].LbaImage.Valid == 0) {
-        break;
-      }
-      ZeroMem (&LoadedExtraImage[Index], sizeof (LOADED_IMAGE));
-      LoadedExtraImage[Index].HwPartHandle  = HwPartHandle;
-      LoadedExtraImage[Index].LoadImageType = LOAD_IMAGE_EXTRA0 + Index;
-      Status = LoadAndSetupImage (&BootOption, &LoadedExtraImage[Index]);
-      if (EFI_ERROR (Status)) {
-        DEBUG ((DEBUG_INFO, "Load extra Image[%d]: Status = %r\n", Index, Status));
-      }
-    }
+  //
+  // Parse Boot Image
+  //
+  Status = ParseBootImages (OsBootOption, LoadedImageHandle);
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_INFO, "Failed to Parse Boot Image\n"));
+    goto ERROR;
   }
 
-  Status = UpdateOsParameters (&BootOption, &LoadedImage, &LoadedTrustyImage, &LoadedExtraImage[0]);
+  //
+  // Setup Boot Image
+  //
+  Status = SetupBootImages (OsBootOption, LoadedImageHandle);
   if (EFI_ERROR (Status)) {
-    DEBUG ((DEBUG_INFO, "UpdateOsParameters: Status = %r\n", Status));
-    return Status;
+    DEBUG ((DEBUG_INFO, "Failed to Setup Boot Image - Status = %r\n", Status));
+    goto ERROR;
   }
-  AddMeasurePoint (0x40E0);
+  AddMeasurePoint (0x40B0);
 
-  if ((BootOption.BootFlags & BOOT_FLAGS_TRUSTY) != 0) {
-    Status = StartBooting (&LoadedTrustyImage);
-  } else {
-    Status = StartBooting (&LoadedImage);
+  //
+  // Update Boot Parameters
+  //
+  Status = UpdateBootParameters (OsBootOption, LoadedImageHandle);
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_INFO, "Failed to Update Boot Parameters - Status = %r\n", Status));
+    goto ERROR;
```

Comment:
Looks we are trying to free all the memory if this boot option boots failure. Inside UpdateBootParameters () I believe there are several places to allocate big memory. we could free them later.

Reply:
Right. This PR is to make general flow in top interface layer. After this PR, we need to improve all memory management parts in sub layers.